src/typings/polymorphic.ts

```diff
@@ -0,0 +1,79 @@
+/* eslint-disable @typescript-eslint/ban-types */
+/*
```

Comment:
Polymorphic types for components.

It’s a pattern for code reuse that’s not nearly as popular as higher-order components, render props, or, of course, custom hooks. But it’s probably the simplest one.

Let me quickly explain how a polymorphic component works. Let’s say we have a shareable `Text` component that allows the caller to configure the font, size and color from a handful of options. Some sample calls would be:

```
<Text font="thin" size="10" color="gray-50">
  Main Heading
</Text>

<Text font="heavy" size="5" color="gray-70">
  Name
</Text>

<Text font="regular" size="4" color="gray-90">
  Body text
</Text>
```

So we’re reusing the same `<Text>` component to display drastically different typographies. And if you care at all about semantic HTML, your first question might be, “well what HTML element are we using to render the text?” And this is exactly the value of polymorphic components. In the first example, we’d likely want to render an `<h1>`. The second should be a `<label>`. And the third, a `<p>`.

We can make `<Text>` “polymorphic” by adding an as prop to it that allows the caller to define the HTML tag to use:

```
<Text as="h1" font="thin" size="10" color="gray-50">
  Main Heading
</Text>

<Text as="label" font="heavy" size="5" color="gray-70">
  Name
</Text>

<Text as="p" font="regular" size="4" color="gray-90">
  Body text
</Text>
```

We are using this principle on tabs.
By default they are of a certain type of element. But, if necessary, they can be changed to another valid HTML element.